### PR TITLE
Completed the PIN Based Authentication in the settings feature!

### LINKED ISSUE #71: Implement PIN based authentication - 1
Currently, we support email+password and fingerprint login. PIN-based authentication will help in devices that don't support fingerprint login, as typing email+password every time would be tedious. 

I have broken down this into two issues since this is a relatively big feature. 

### 1. Add "Enable PIN-based login" `SwitchListTile` in security settings

We store all user-based settings in `UserConfigModel`, add a new field `isPINLoginEnabled` to store the value of `SwtichListTile`

When user enables this switch we need to check if `userId` (present in `UserConfigCubit`) is not equal to `guest_user_id`. Becuase if guest users forget their PIN, there is no way to recover the data (Same check is also added to "Enable Fingeprint Login" `SwitchListTile`).

### 2. Display a popup like this to collect 4-digit PIN

Keyboard should only display numbers, add validation to check if both PINs are matching, otherwise display error in a error message. Refer `passwordResetPopup`, most parts of this widget can be re-used. 

### 3, Saving the PIN

Create a new repository `PINAuthRepository` in `lib/features/auth/data/repositories` and implement `savePIN` in `PINAuthRepository` and store it in [flutter_secure_storage](https://pub.dev/packages/flutter_secure_storage) in the form of key:value pairs, Include `userId` in key (eg: <userId>_PIN) so that it PIN's of two user's don't clash.

### PR DESCRIPTION
**Pull request by LujainAlMansoori**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
The PIN-based authentication feature is complete in the settings. The user is able to select to log in through a pin from the setting because we added a Enable PIN-based login. Once this takes place, a pop up comes up and collects the 4 digit pin from the user. We added a validation where only keyword numbers show up and we check if both PINs are matching, otherwise display error in a error message. Once that happens the user id and hashed pin is saved into a PINAuthRepository in the form of key:value pairs. This also resolves task 2 where the user can login through their userId if their userId is found in the flutter_storage, to which the userId is found with an associated hashed pin. The pin_login_form.dart involves utilizing an event from auth_form_bloc to ensure the user's pin is authenticated through pin_auth_repository.dart.

## Featured Covered in this PR

- [X] Task 1
- [X] Task 2

## Related Tickets & Documents

- Related Issue #71 and #72
- Closes #71 and #72

## Screenshots, Recordings
Screenshot 2023-11-29 at 8 42 13 PM
Screenshot 2023-11-29 at 8 42 21 PM
Screenshot 2023-11-29 at 8 42 28 PM
image
image

This image shows when the pin is not the same as the user's saved pin:
image

## Tested Feature??

- [ ] In Real Device.
- [X] In Emulator

### CODE REVIEW

**Review by mbfannan — APPROVED** [review]
Looks great, wonderful work all contributers!

**Comment by SankethBK** [comment]
Thanks for contributing! I'll review by tomorrow

**Comment by LujainAlMansoori** [comment]
Sure, thanks!

**Comment by mbfannan** [comment]
@SankethBK any updates?

File: lib/features/auth/presentation/widgets/security_settings.dart [hunk 1]
```diff
@@ -10,201 +10,266 @@ import 'package:dairy_app/features/auth/presentation/bloc/user_config/user_confi
...
+                             return;
+                           }
+                           if (value) {
+                             // Call the PIN reset popup
+                             await pinResetPopup(
+                               context: context,
+                               userPinId: userId,
+                             );
+                           }
+
```
Comment (SankethBK) [inline lib/features/auth/presentation/widgets/security_settings.dart:251]: This switch should be enabled only if PIN is successfully set, currently it gets enabled if we enter mismatched pins

**Review by SankethBK — CHANGES_REQUESTED** [review]
Nice job! I have added minor comments, check them once

**Review by mbfannan — APPROVED** [review]
Thanks for the changes!

**Comment by LujainAlMansoori** [comment]
Thanks for your changes, I resolved them!

**Comment by mbfannan** [comment]
Hi @SankethBK and @LujainAlMansoori took me a while because some formatting and syntax errors didn't appear before, but everything should be alright now to merge thanks! Once it's merged I'll get on to issue #72 and I already got a start on it thanks!

**Review by mbfannan — COMMENTED** [review]
With the additional changes made hopefully everything should be fine, just to note however another push was made to change file name of "PINAuthRepository.dart" due to linter error along with other formatting issues resolved.

File: lib/features/auth/presentation/widgets/security_settings.dart [hunk 2]
```diff
@@ -196,6 +198,49 @@ class SecuritySettings extends StatelessWidget {
...
+                              showToast(S.current.unexpectedErrorOccured);
+                              return;
+                            }
+
+                            if (userId == GuestUserDetails.guestUserId) {
+                              showToast(S.current
+                                  .pleaseSetupYourAccountToUseThisFeature);
+                              return;
+                            }
+                            if (value) {
```
Comment (SankethBK) [inline lib/features/auth/presentation/widgets/security_settings.dart:225]: Use value == true, improves readability

File: lib/features/auth/presentation/widgets/security_settings.dart [hunk 3]
```diff
@@ -196,6 +198,49 @@ class SecuritySettings extends StatelessWidget {
...
+                                context: context,
+                                userPinId: userId,
+                              );
+                            }
+
+                            try {
+                              userConfigCubit.setUserConfig(
+                                UserConfigConstants.isPINLoginEnabled, //change
+                                value,
+                              );
```
Comment (SankethBK) [inline lib/features/auth/presentation/widgets/security_settings.dart:237]: Currently the switch gets enabled even if pin is not successfully set, you can return a boolean value from `pinResetPopup` using `Navigator.pop(context)` and enable this switch only if its true. You can use `passwordLoginPopup` as reference for returning data from `Navigator.pop(context)`

File: pubspec.lock [hunk 4]
```diff
@@ -1,1903 +0,0 @@
-# Generated by pub
-# See https://dart.dev/tools/pub/glossary#lockfile
-packages:
```
Comment (SankethBK) [inline pubspec.lock:3]: We need this file, don't delete it

**Comment by SankethBK** [comment]
Hey @mbfannan , I have added minor comments, I have changed the target branch to `pin_authentication` as we need to have part 2 as well before releasing it. Feel free to merge this to `pin_authentication` branch, you can resolve the comments later. 

Thanks @mbfannan @LujainAlMansoori for your contribution!

**Comment by mbfannan** [comment]
Hi @SankethBK, @LujainAlMansoori  and I resolved task 1 for issue #71 and completed issue #72. Hope everything's alright to merge!

**Comment by LujainAlMansoori** [comment]
Yes, we completed both issues, issues #71 and #72  and they should be ready to be merged!

**Comment by mbfannan** [comment]
Hi, @SankethBK do you have any updates? We're doing this as part of a Software Engineering course at Carnegie Mellon University, and our deadline to continue contributing would be at the latest by tomorrow evening. So please let us know if we can help with any more changes to merge, thanks!

**Comment by SankethBK** [comment]
Hey @mbfannan I'll review in next 2-3 hours.

**Comment by mbfannan** [comment]
> Hey @mbfannan I'll review in next 2-3 hours.

Great thanks!

**Comment by SankethBK** [comment]
This is great! thanks @mbfannan @LujainAlMansoori for choosing this project to contribute.

**Pull request merged**